papers/pydra/paper.rst

```diff
@@ -0,0 +1,965 @@
+
+.. figure:: classes.pdf
+   :figclass: h!
+   :scale: 30%
+
+   A schematic presentation of principal classes in Pydra. :label:`classes`
+
+Dataflows Components: Task and Workflow
+=======================================
+A *Task* is the basic runnable component of *Pydra* and is described by the
+class ``TaskBase``. A *Task* has named inputs and outputs thus allowing
+construction of dataflows. It can be hashed and executes in a specific working
+directory. Any Pydra *Task* can be used as a function in a script, thus allowing
+dual use in Pydra *Workflows* and in standalone scripts. There are several
+classes that inherit from ``TaskBase`` and each has a different application:
+
+* ``FunctionTask`` is a *Task* that executes Python functions. Most Python functions
+  declared in an existing library, package, or interactively in a terminal can
+  be converted to a ``FunctionTask`` by using *Pydra* decorator - ``mark.task``.
+
+  .. code-block:: python
+
+     import numpy as np
+     from pydra import mark
+     fft = mark.annotate({'a': np.ndarray,
+                      'return': float})(np.fft.fft)
+     fft_task = mark.task(fft)()
+     result = fft_task(a=np.random.rand(512))
+
+
+  `fft_task` is now a *Pydra* task and result will contain a *Pydra*'s ``Result`` object.
+  In addition, the user can use Python's function annotation or another *Pydra*
+  decorator |---| ``mark.annotate`` in order to specify the output. In the
+  following example, we decorate an arbitrary Python function to create named
+  outputs:
+
+  .. code-block:: python
+
+     @mark.task
+     @mark.annotate(
+         {"return": {"mean": float, "std": float}}
+     )
+     def mean_dev(my_data):
+         import statistics as st
+         return st.mean(my_data), st.stdev(my_data)
+
+     result = mean_dev(my_data=[...])()
+
+  When the task is executed `result.output` will contain two attributes: `mean`
+  and `std`. These named outputs allowing passing different outputs to
+  different downstream nodes in a dataflow.
+
+* ``ShellCommandTask`` is a *Task* used to run shell commands and executables.
+  It can be used with a simple command without any arguments, or with specific
+  set of arguments and flags, e.g.:
+
+  .. code-block:: python
+
+     ShellCommandTask(executable="pwd")
+
+     ShellCommandTask(executable="ls", args="my_dir")
+
+  The *Task* can accommodate more complex shell commands by allowing the user to
+  customize inputs and outputs of the commands.
+  One can generate an input
+  specification to specify names of inputs, positions in the command, types of
+  the inputs, and other metadata.
+  As a specific example, FSL's BET command (Brain
+  Extraction Tool) can be called on the command line as:
+
+  .. code-block:: python
+
+    bet input_file output_file -m
+
+  Each of these inputs can be augmented as a named argument to the
+  ``ShellCommandTask``. As shown next, even an output is specified by constructing
```

Comment:
Tego ostatniego zdania nie rozumiem.

Reply:
ta, to satra mi zmienil, pomysle nad nim